crates/sparrow-runtime/src/read/stream_reader.rs

```diff
@@ -0,0 +1,148 @@
+use std::sync::Arc;
+
+use anyhow::Context;
+use arrow::datatypes::SchemaRef;
+use arrow::record_batch::RecordBatch;
+use chrono::NaiveDateTime;
+use error_stack::{IntoReport, IntoReportCompat, ResultExt};
+use futures::stream::BoxStream;
+use futures::{Stream, StreamExt};
+use hashbrown::HashSet;
+use itertools::Itertools;
+use sparrow_api::kaskada::v1alpha::slice_plan::Slice;
+use sparrow_api::kaskada::v1alpha::{PulsarSource, PulsarSubscription};
+use sparrow_compiler::{DataContext, TableInfo};
+use sparrow_core::TableSchema;
+use sparrow_qfr::{
+    activity, gauge, Activity, FlightRecorder, Gauge, PushRegistration, Registration, Registrations,
+};
+use tracing::info;
+
+use crate::data_manager::{DataHandle, DataManager};
+use crate::execute::operation::OperationContext;
+use crate::merge::{homogeneous_merge, GatheredBatches, Gatherer};
+use crate::min_heap::{HasPriority, MinHeap};
+use crate::read::error::Error;
+use crate::read::parquet_stream::{self, new_parquet_stream};
+use crate::streams::pulsar::stream;
+use crate::{prepare, streams, Batch, RawMetadata};
+
+const READ_STREAM: Activity = activity!("scan.read_stream");
+
+const MIN_BATCH_TIME: Gauge<i64> = gauge!("min_time_in_batch");
+const MAX_BATCH_TIME: Gauge<i64> = gauge!("max_time_in_batch");
+const NUM_INPUT_ROWS: Gauge<usize> = gauge!("num_input_rows");
+const NUM_OUTPUT_ROWS: Gauge<usize> = gauge!("num_output_rows");
+
+static REGISTRATION: Registration = Registration::new(|| {
+    let mut r = Registrations::default();
+    r.add(READ_STREAM);
+
+    r.add(MIN_BATCH_TIME);
+    r.add(MAX_BATCH_TIME);
+    r.add(NUM_INPUT_ROWS);
+    r.add(NUM_OUTPUT_ROWS);
+    r
+});
+
+inventory::submit!(&REGISTRATION);
+
+/// The bounded lateness parameter, which configures the delay in watermark time.
+///
+/// In practical terms, this allows for items in the stream to be within 1 second
+/// compared to the max timestamp read.
+///
+/// This is hard-coded for now, but could easily be made configurable as a parameter
+/// to the table. This simple hueristic is a good start, but we can improve on this
+/// by statistically modeling event behavior and adapting the watermark accordingly.
+const BOUNDED_LATENESS_NS: i64 = 1_000_000_000;
+
+/// Create a stream that continually reads messages from a stream.
+pub(crate) async fn stream_reader(
+    context: &OperationContext,
+    table_info: &TableInfo,
+    requested_slice: Option<&Slice>,
+    projected_columns: Option<Vec<String>>,
+    _flight_recorder: FlightRecorder,
+    pulsar_source: &PulsarSource,
+) -> error_stack::Result<impl Stream<Item = error_stack::Result<Batch, Error>> + 'static, Error> {
+    let pulsar_subscription =
+        std::env::var("PULSAR_SUBSCRIPTION").unwrap_or("subscription-default".to_owned());
```

Comment:
we might want the `subscription-id` to be the `materialization-id` +`query-hash` (or something similar) when running from wren.

Reply:
Yeah, this is something I need to check out as well, in regards to restarting reading with a specific subscription. I'll add a TODO about that here.